### PR TITLE
track TCP flags in first packet in each direction

### PR DESCRIPTION
**Pull request by rtkjweeks**
Suricata saves the TCP flags which it sees inside
`struct TcpStream_::tcp_flags` but this contains a union of
all flags seen.  It can be beneficial to know the flags
used solely in the first packet seen (in each direction).

This change stores the initial flags seen in each direction,
even if the first packet seen is a FIN or a RST (if midstream
is configured)

Make sure these boxes are signed before submitting your Pull Request -- thank you.

- [x] I have read the contributing guide lines at https://redmine.openinfosecfoundation.org/projects/suricata/wiki/Contributing
- [x] I have signed the Open Information Security Foundation contribution agreement at https://suricata-ids.org/about/contribution-agreement/
- [ ] I have updated the user guide (in doc/userguide/) to reflect the changes made (if applicable)

Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket:

Describe changes:
-
-
-

suricata-verify-pr: 432
#suricata-verify-repo:
#suricata-verify-branch:
#suricata-update-pr:
#suricata-update-repo:
#suricata-update-branch:
#libhtp-pr:
#libhtp-repo:
#libhtp-branch:

### CODE REVIEW

File: src/stream-tcp.c [hunk 1]
```diff
@@ -908,14 +910,54 @@ static int StreamTcpPacketStateNone(ThreadVars *tv, Packet *p,
...
     if (p->tcph->th_flags & TH_RST) {
-        StreamTcpSetEvent(p, STREAM_RST_BUT_NO_SESSION);
-        SCLogDebug("RST packet received, no session setup");
-        return -1;
+        if (stream_config.midstream == FALSE) {
+            StreamTcpSetEvent(p, STREAM_RST_BUT_NO_SESSION);
+            SCLogDebug("RST packet received, no session setup");
+            return -1;
+        }
+        ssn = StreamTcpNewSession(p, stt->ssn_pool_id);
```
Comment (victorjulien) [inline src/stream-tcp.c:918]: these changes should be in a separate commit, as they are changing Suricata's midstream behavior and are not just about flags logging

Comment (victorjulien) [reply src/stream-tcp.c:918]: you'll have to convince me that this makes sense at all btw, it seems setting up a tcp session based on a RST packet makes no sense

File: src/stream-tcp.c [hunk 2]
```diff
@@ -4792,10 +4834,15 @@ int StreamTcpPacket (ThreadVars *tv, Packet *p, StreamTcpThread *stt,
     /* track TCP flags */
     if (ssn != NULL) {
         ssn->tcp_packet_flags |= p->tcph->th_flags;
-        if (PKT_IS_TOSERVER(p))
+        if (PKT_IS_TOSERVER(p)) {
             ssn->client.tcp_flags |= p->tcph->th_flags;
-        else if (PKT_IS_TOCLIENT(p))
+            if (unlikely(ssn->client.tcp_init_flags == 0))
```
Comment (victorjulien) [inline src/stream-tcp.c:4839]: how can this happen? If we have a ssn we should have run the code first setting the flags?

**Review by victorjulien — CHANGES_REQUESTED** [review]
Needs a look at the CI failures, looks like there is more than just the formatting issue.

Git id looks off.

File: src/stream-tcp.c (hunk 1 above)
Comment (rtkjweeks) [reply src/stream-tcp.c:918]: The use case here is in anomaly detection; in sudden increase in flows containing only FIN or RST packets would be an indication of strange TCP behaviour, and possibly something that would need further indication.

I could probably make it configurable.

File: src/stream-tcp.c (hunk 2 above)
Comment (rtkjweeks) [reply src/stream-tcp.c:4839]: If the first packet seen was TOSERVER then we would've initialized that direction, but not this direction (Technically only relevant if https://github.com/OISF/suricata/pull/5802 is accepted, I guess)

**Comment by rtkjweeks** [comment]
Not quite sure how to investigate the integration failures here... they pass on some systems, but not others.  It's the same test failing, but appears to be environment related and, unfortunately, the tests all pass for me in my environment.

I'll see if I can find some different systems to test on, in the meantime, but any suggestions here would be greatly appreciated.

**Comment by catenacyber** [comment]
> Not quite sure how to investigate the integration failures here...

You should investigate the _first_ failing test

**Comment by rtkjweeks** [comment]
Closing per V2 PR: https://github.com/OISF/suricata/pull/5843
